### ISSUE TITLE
Templates in client_definitions

### ISSUE THREAD

**Issue by michaelpporter** [cmacrae/ansible-sensu#211]
https://ansible-sensu.readthedocs.io/en/latest/dynamic_checks/

The docs show templates in the `client_definitions` folder. I have 

```
$ tree data/static
data/static
`-- sensu
    |-- client_definitions
    |   |-- sensu_rabbitmq_servers
    |   |   `-- check_users.json.j2
    |   `-- webservers
    |       `-- check_uptime.json.j2
```

The loop for client_definitions [copies](https://docs.ansible.com/ansible/latest/modules/copy_module.html) the directory over, not individual files. [template](https://docs.ansible.com/ansible/latest/modules/template_module.html) is needed if you want to us` j2` templates. But, `template` is not supported in this context.

```yaml
- name: Deploy client definitions
  copy:
    src: "{{ static_data_store }}/sensu/client_definitions/{{ item }}/"
    dest: "{{ sensu_config_path }}/conf.d/{{ item | basename | regex_replace('.j2', '') }}"
    owner: "{{ sensu_user_name }}"
    group: "{{ sensu_group_name }}"
  when:
    - sensu_available_client_definitions is defined
    - sensu_available_checks is not skipped
    - item in sensu_available_client_definitions.stdout_lines
  loop: "{{ group_names|flatten }}"
  notify: restart sensu-client service
```

I myself have put static checks in the `client_definitions` for now and used [Check tokens](https://docs.sensu.io/sensu-core/1.1/reference/checks/#reference-documentation) in a custom `client.json.j2` template

**client.json.j2**
```j2
{
  "client": {
    "name": "{{ sensu_client_name }}",
    "address": "{{ ansible_default_ipv4['address'] }}",
    "subscriptions": {{ sensu_client_subscriptions | to_nice_json(indent=6) }},
    {% if sensu_client_slack_channel %}
    "slack_channel": "{{ slack_channel }}",
    {% endif %}
    "keepalive": {
      "handlers": {{ sensu_client_keepalive_handlers | to_nice_json(indent=8) }},
      "thresholds": {
        "warning": {{ sensu_client_keepalive_threshold_warning }},
        "critical": {{ sensu_client_keepalive_threshold_critical }}
        }
    },
    "global": {
        "domain": "{{ checks_global_domain }}",
        "check_disk_flags": "{{ sensu_disk_mount }}"
    },
    "es_node": {
        "user": "{{ checks_es_node_user }}",
        "password": "{{ checks_es_node_password }}"
    },
    "mysql": {
        "user": "{{ checks_mysql_user }}",
        "password": "{{ checks_mysql_password }}",
        "sock": "{{ checks_mysql_sock }}"
    },
    "redact": ["password", "user"],
    "safe_mode": {{ sensu_client_safe_mode | bool | lower }},
    "environment": "{{ checks_environment }}"
  }
}

```

**client_definitions/metrics_es_cluster.json**
```json
{
    "checks": {
        "es_cluster_metrics": {
            "command": "metrics-es-cluster.rb -h :::address::: -u :::es_node.user::: -P :::es_node.password::: -a -s elasticsearch",
            "subscribers": [
                "roundrobin:es"
            ],
            "ttl": 1800,
            "source": "es-cluster",
            "interval": 60,
            "timeout": 30,
            "refresh": 1800,
            "handler": "relay",
            "slack_channel": "#dev-alerts",
            "collection": "elastic_cluster",
            "type": "metric"
        }
    }
}
```

This might be easier to update the docs then convert to template use.

**Comment by jaredledvina** [comment]
Hi again @michaelpporter,

This is a good catch,  as far as I can tell, that logic was never used template: https://github.com/sensu/sensu-ansible/commit/488a9a727bac913a9f2744ca297156b4d8c599c3#diff-2be5f2ebe724642f1c79fbf55ef84f47R98 

I went back through the issue/PR and reviewed https://github.com/sensu/sensu-ansible/commit/bb18e42a0cf5afd9865702ff7c04fee534a52b74, I think that this is a documentation issue and that it was never intended to actually template out Jinja files. So, that all  said, using check token substitution is a great way to work around this role not support native Jinja templates for the `client_definitions`. 

I'd be willing to see this support native Jinja templates though as well. Unless I haven't had enough coffee today, I _think_ we just need to change https://github.com/sensu/sensu-ansible/blob/45b93849fa162936f21f82032de6ca4dd4dca174/tasks/plugins.yml#L105 to `template`.

**Comment by michaelpporter** [comment]
I tested template with files and it did not work. It fails on the folder as it’s not a template file. The TravisCI tests does not have files so the task is skipped.

**Comment by michaelpporter** [comment]
FYI, the output using template:

```bash
TASK [external/sensu-ansible : Deploy client definitions] ***********************************************************************************
failed: [sensu-checks.######.com] (item=sensu_checks) => {"changed": false, "item": "sensu_checks", "msg": "could not find src=/Users/michaelporter/######/awx-plays/plays/data/static/sensu/client_definitions/sensu_checks, Could not find or access '/Users/michaelporter/######/awx-plays/plays/data/static/sensu/client_definitions/sensu_checks' on the Ansible Controller.\nIf you are using a module and expect the file to exist on the remote, see the remote_src option"}
failed: [sensu-checks.######.com] (item=sensu_clients) => {"changed": false, "item": "sensu_clients", "msg": "could not find src=/Users/michaelporter/######/awx-plays/plays/data/static/sensu/client_definitions/sensu_clients, Could not find or access '/Users/michaelporter/######/awx-plays/plays/data/static/sensu/client_definitions/sensu_clients' on the Ansible Controller.\nIf you are using a module and expect the file to exist on the remote, see the remote_src option"}
```

Partial listing of that folder:

```bash
± |master U:2 ✗| → ll /Users/michaelporter/######/awx-plays/plays/data/static/sensu/client_definitions/sensu_checks
total 296
drwxr-xr-x  39 michaelporter  staff  1248 Feb 13 06:12 ./
drwxr-xr-x  10 michaelporter  staff   320 Feb 23 09:46 ../
-rw-r--r--@  1 michaelporter  staff   565 Feb 13 11:51 check_elasticsearch_snapshot.json
```


Output as `copy`:

```bash

TASK [external/sensu-ansible : Deploy client definitions] ***********************************************************************************
ok: [sensu-checks.######.com] => (item=sensu_checks)
ok: [sensu-checks.######.com] => (item=sensu_clients)
```

**Comment by michaelpporter** [comment]
This could become a new function

```yaml
- name: Ensure local directories exist
  file:
    state: directory
    dest: "{{ static_data_store }}/sensu/{{ item }}"
  delegate_to: localhost
  become: no
  loop:
    - checks
    - filters
    - handlers
    - mutators
    - definitions
    - client_definitions
    - client_templates # add templates
``` 

```yaml
- name: Register available client templates
  command: "ls {{ static_data_store }}/sensu/client_templates"
  delegate_to: localhost
  register: sensu_available_client_templates
  changed_when: false
  become: false

- name: Ensure directory structure exists
  file:
    path: '{{ sensu_config_path }}/conf.d/{{ item.path | dirname }}'
    state: directory
  with_filetree: '{{ static_data_store }}/sensu/client_templates'
  when:
    - item.path | dirname in group_names
    - sensu_available_client_templates is defined
    - sensu_available_client_templates is not skipped

- name: Deploy client templates
  template:
    src: "{{ static_data_store }}/sensu/client_templates/{{ item }}/"
    dest: "{{ sensu_config_path }}/conf.d/{{ item.path | dirname }}/{{ item.path | basename | regex_replace('.j2', '') }}"
    owner: "{{ sensu_user_name }}"
    group: "{{ sensu_group_name }}"
  with_filetree: "{{ static_data_store }}/sensu/client_templates"
  when:
    - item.path | dirname in group_names
    - sensu_available_client_templates is defined
    - sensu_available_client_templates is not skipped
  notify: restart sensu-client service
```

edited `copy` to `template`

**Comment by jaredledvina** [comment]
Ah @michaelpporter, good catch! I was going to test that PR some more locally (had to run some errands after opening it). So, with that,  perhaps we do just move templates into their own setup. 

Perhaps, I'm just missing it, but in your example above, where do the `.j2` templates get templated out? Or are you saying that's the task we should right? I think that seems like a totally fair addition in general though.

**Comment by michaelpporter** [comment]
@jaredledvina the last one should be template, not copy. I edited it. I have not tested it with live templates. I did run though it with `debug. I can do more testing tomorrow or Monday.

**Comment by michaelpporter** [comment]
PS: I got the idea and basic structure from https://stackoverflow.com/questions/41667864/can-the-templates-module-handle-multiple-templates-directories

**Comment by michaelpporter** [comment]
PSS: Because `with_filetree` loops over file. if you have a folder with 5 templates it will try to make the directory 5 times. There might need to be a step to build the array so the remote connections are limited.

**Comment by michaelpporter** [comment]
This seems to work.

```yaml
- name: Register available client templates
  command: "ls {{ static_data_store }}/sensu/client_templates"
  delegate_to: localhost
  register: sensu_available_client_templates
  changed_when: false
  become: false

- name: Deploy client template folders
  file:
    path: '{{ sensu_config_path }}/conf.d/{{ item | basename }}'
    state: directory
    owner: "{{ sensu_user_name }}"
    group: "{{ sensu_group_name }}"
  when:
    - sensu_available_client_templates is defined
    - sensu_available_client_templates is not skipped
    - item in sensu_available_client_templates.stdout_lines
  loop: "{{ group_names|flatten }}"
  notify: restart sensu-client service

- name: Deploy client templates
  template:
    src: "{{ static_data_store }}/sensu/client_templates/{{ item.path | dirname }}/{{ item.path | basename }}"
    dest: "{{ sensu_config_path }}/conf.d/{{ item.path | dirname }}/{{ item.path | basename | regex_replace('.j2', '') }}"
    owner: "{{ sensu_user_name }}"
    group: "{{ sensu_group_name }}"
  with_filetree: "{{ static_data_store }}/sensu/client_templates"
  when:
    - item.state == 'file'
    - item.path | dirname in group_names
  notify: restart sensu-client service
```

Locally I updated molecule for Ubuntu. 14.04 is failing the inspec tests on the ports.

```yaml
platforms:
  - name: ubuntu-14.04
    image: dokken/ubuntu-14.04
    command: /sbin/init
    capabilities:
      - SYS_ADMIN
    groups:
      - sensu_checks
  - name: ubuntu-16.04
    image: dokken/ubuntu-16.04
    command: /bin/systemd
    capabilities:
      - SYS_ADMIN
    groups:
      - sensu_checks
    volumes:
      - /sys/fs/cgroup:/sys/fs/cgroup:ro
  - name: ubuntu-18.04
    image: dokken/ubuntu-18.04
    command: /bin/systemd
    capabilities:
      - SYS_ADMIN
    groups:
      - sensu_checks
    volumes:
      - /sys/fs/cgroup:/sys/fs/cgroup:ro
# .....
  inventory:
    group_vars:
      all:
        sensu_master: true
        sensu_include_dashboard: true
        sensu_rabbitmq_server: true
        sensu_redis_server: true
        sensu_rabbitmq_host: "{{ ansible_hostname }}"
        sensu_redis_host: "{{ ansible_hostname }}"
        sensu_api_host: "{{ ansible_hostname }}"
        ansible_default_ipv4:
          address: 127.0.0.1
        sensu_remote_plugins:
          - sensu-plugins-disk-checks
        sensu_check_interval: 60
```

```rb
# Ensure disk check exists
describe file('/etc/sensu/conf.d/sensu_masters/check_disk_usage.json') do
  it { should exist }
  its('content') { should match(/check-disk-usage/) }
end

# Ensure disk metrics exists
describe file('/etc/sensu/conf.d/sensu_checks/metrics_disk_usage.json') do
  it { should exist }
  its('content') { should match(/metrics-disk-usage/) }
  its('content') { should match(/60/) }
end

# Ensure not_used does not exist
describe file('/etc/sensu/conf.d/not_used/not_a_check.json') do
  it { should_not exist }
end
```

```bash
tree
.
|____sensu
| |____filters
| |____client_templates
| | |____not_used
| | | |____not_a_check.json.j2
| | |____sensu_checks
| | | |____metrics_disk_usage.json.j2
| |____checks
| |____client_definitions
| | |____sensu_masters
| | | |____check_disk_usage.json
| |____mutators
| |____definitions
| |____handlers
```

**Comment by jaredledvina** [comment]
PR merged and released in 5.1.0! 
https://github.com/sensu/sensu-ansible/releases/tag/5.1.0

Thanks!!
